feat: add C `ndarray` API and refactor `blas/ext/base/dsort2hp` (#3292)
stdlib-js/stdlib: lib/node_modules/@stdlib/blas/ext/base/dsort2hp/README.md

-   **N**: number of indexed elements.
-   **order**: sort order. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.
-   **x**: first input [`Float64Array`][@stdlib/array/float64].

> -   **strideX**: `x` stride length.

Author: ```suggestion
-   **strideX**: stride length for `x`.
```

-   **y**: second input [`Float64Array`][@stdlib/array/float64].

> -   **strideY**: `y` stride length.

Author: ```suggestion
-   **strideY**: stride length for `y`.
```

> The `N` and stride parameters determine which elements in `x` and `y` are accessed at runtime. For example, to sort every other element:

Author: ```suggestion
The `N` and stride parameters determine which elements in the strided arrays are accessed at runtime. For example, to sort every other element:
```

-   **offsetX**: `x` starting index.

> -   **offsetY**: `y` starting index.

Author: ```suggestion
-   **offsetX**: starting index for `x`.
-   **offsetY**: starting index for `y`.
```

<!-- /.examples -->

<!-- C interface documentation. -->

* * *

<section class="c">

## C APIs

<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->

<section class="intro">

</section>

<!-- /.intro -->

<!-- C usage documentation. -->

<section class="usage">

### Usage

```c
#include "stdlib/blas/ext/base/dsort2hp.h"
```

#### stdlib_strided_dsort2hp( N, order, \*X, strideX, \*Y, strideY )

Simultaneously sorts two double-precision floating-point strided arrays based on the sort order of the first array using heapsort.

```c
double x[] = { 1.0, -2.0, 3.0, -4.0 };
double y[] = { 0.0, 1.0, 2.0, 3.0 };

> stdlib_strided_dsort2hp( 4, 1, x, 1, y, 1 );

Author: ```suggestion
stdlib_strided_dsort2hp( 4, 1.0, x, 1, y, 1 );
```

```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.

> -   **order**: `[in] CBLAS_INT` sort order. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.

Author: ```suggestion
-   **order**: `[in] double` sort order. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.
```

-   **X**: `[inout] double*` input array.
-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
-   **Y**: `[inout] double*` input array.
-   **strideY**: `[in] CBLAS_INT` stride length for `Y`.

```c

> stdlib_strided_dsort2hp( const CBLAS_INT N, const CBLAS_INT order, double *X, CBLAS_INT strideX, double *Y, CBLAS_INT strideY );

Author: ```suggestion
stdlib_strided_dsort2hp( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX, double *Y, CBLAS_INT strideY );
```

```

<!--lint disable maximum-heading-length-->

#### stdlib_strided_dsort2hp_ndarray( N, order, \*X, strideX, offsetX, \*Y, strideY, offsetY )

<!--lint enable maximum-heading-length-->

Simultaneously sorts two double-precision floating-point strided arrays based on the sort order of the first array using heapsort and alternative indexing semantics.

```c
double x[] = { 1.0, -2.0, 3.0, -4.0 };
double y[] = { 0.0, 1.0, 2.0, 3.0 };

> stdlib_strided_dsort2hp_ndarray( 4, 1, x, 1, 0, y, 1, 0 );

Author: ```suggestion
stdlib_strided_dsort2hp_ndarray( 4, 1.0, x, 1, 0, y, 1, 0 );
```

```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.

> -   **order**: `[in] double` sort order.

Author: ```suggestion
-   **order**: `[in] double` sort order. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged..
```

> -   **order**: `[in] CBLAS_INT` sort order.

Author: ```suggestion
-   **order**: `[in] double` sort order.
```

> -   **X**: `[inout] double*` input array. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.

Author: ```suggestion
-   **X**: `[inout] double*` input array.
```

-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
-   **offsetX**: `[in] CBLAS_INT` starting index for `X`.
-   **Y**: `[inout] double*` input array.
-   **strideY**: `[in] CBLAS_INT` stride length for `Y`.
-   **offsetY**: `[in] CBLAS_INT` starting index for `Y`.

```c

> stdlib_strided_dsort2hp_ndarray( const CBLAS_INT N, const CBLAS_INT order, double *X, CBLAS_INT strideX, CBLAS_INT offsetX, double *Y, CBLAS_INT strideY, CBLAS_INT offsetY );

Author: ```suggestion
stdlib_strided_dsort2hp_ndarray( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX, CBLAS_INT offsetX, double *Y, CBLAS_INT strideY, CBLAS_INT offsetY );
```

<section class="intro">

</section>

<!-- /.intro -->

<!-- C usage documentation. -->

<section class="usage">

### Usage

```c
#include "stdlib/blas/ext/base/dsort2hp.h"
```

#### stdlib_strided_dsort2hp( N, order, \*X, strideX, \*Y, strideY )

Simultaneously sorts two double-precision floating-point strided arrays based on the sort order of the first array using heapsort.

```c
double x[] = { 1.0, -2.0, 3.0, -4.0 };
double y[] = { 0.0, 1.0, 2.0, 3.0 };

stdlib_strided_dsort2hp( 4, 1, x, 1, y, 1 );
```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **order**: `[in] CBLAS_INT` sort order. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.
-   **X**: `[inout] double*` input array.
-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
-   **Y**: `[inout] double*` input array.
-   **strideY**: `[in] CBLAS_INT` stride length for `Y`.

```c

> stdlib_strided_dsort2hp( const CBLAS_INT N, const CBLAS_INT order, double *X, const CBLAS_INT strideX, double *Y, const CBLAS_INT strideY );

Author: ```suggestion
stdlib_strided_dsort2hp( const CBLAS_INT N, const CBLAS_INT order, double *X, CBLAS_INT strideX, double *Y, CBLAS_INT strideY );
```

```

<!--lint disable maximum-heading-length-->

#### stdlib_strided_dsort2hp_ndarray( N, order, \*X, strideX, offsetX, \*Y, strideY, offsetY )

<!--lint enable maximum-heading-length-->

Simultaneously sorts two double-precision floating-point strided arrays based on the sort order of the first array using heapsort and alternative indexing semantics.

```c
double x[] = { 1.0, -2.0, 3.0, -4.0 };
double y[] = { 0.0, 1.0, 2.0, 3.0 };

stdlib_strided_dsort2hp_ndarray( 4, 1, x, 1, 0, y, 1, 0 );
```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **order**: `[in] CBLAS_INT` sort order.
-   **X**: `[inout] double*` input array. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.
-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
-   **offsetX**: `[in] CBLAS_INT` starting index for `X`.
-   **Y**: `[inout] double*` input array.
-   **strideY**: `[in] CBLAS_INT` stride length for `Y`.
-   **offsetY**: `[in] CBLAS_INT` starting index for `Y`.

```c

> stdlib_strided_dsort2hp_ndarray( const CBLAS_INT N, const CBLAS_INT order, double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX, double *Y, const CBLAS_INT strideY, const CBLAS_INT offsetY );

Author: ```suggestion
stdlib_strided_dsort2hp_ndarray( const CBLAS_INT N, const CBLAS_INT order, double *X, CBLAS_INT strideX, CBLAS_INT offsetX, double *Y, CBLAS_INT strideY, CBLAS_INT offsetY );
```
